model/publishing/delivery/RemoteDeliveryPublisher.php

```diff
@@ -0,0 +1,199 @@
+                $requestData[] = [
+                    'name' => RestTest::REST_DELIVERY_CLASS_LABEL,
+                    'contents' => $deliveryClass->getLabel()
+                ];
+            }
+
+            return $requestData;
+        } catch (FileNotFoundException $e) {
+            $this->logError($e->getMessage(), [$e->__toString()]);
+            $message = sprintf(__('QTI Test backup file not found for delivery "%s"'), $this->delivery->getLabel());
+            throw new PublishingFailedException($message);
+        }
+    }
+
+    /**
+     * @param core_kernel_classes_Resource $delivery
+     * @return File
+     * @throws core_kernel_persistence_Exception
+     */
+    private function getQtiTestPackageFile(core_kernel_classes_Resource $delivery): File
+    {
+        $packagePath = $delivery->getOnePropertyValue($this->getProperty(TestBackupService::PROPERTY_QTI_TEST_BACKUP_PATH));
+        $packagePath = (string)  $packagePath;
+
+        return $this->getServiceLocator()
+            ->get(FileSystemService::SERVICE_ID)
+            ->getDirectory(TestBackupService::FILESYSTEM_ID)
+            ->getFile($packagePath);
+    }
+
+    /**
+     * @param array $requestData
+     * @return ResponseInterface
+     */
+    private function callRemotePublishingApi(array $requestData): ResponseInterface
+    {
+        try {
+            $body = new MultipartStream($requestData);
+            $request = new Request('POST', '/taoDeliveryRdf/RestTest/compileDeferred');
+            $request = $request->withBody($body);
+
+            return $this->getPlatformService()->callApi($this->environment->getUri(), $request);
+        } catch (ConnectException $e) {
+            $this->logError($e->getMessage(), [$e->__toString()]);
+
+            $message = __('Remote environment "%s" is not reachable.', $this->environment->getLabel());
+            throw new PublishingFailedException($message);
+        } catch (ClientException $e) {
+            $this->logError($e->getMessage(), [$e->__toString()]);
+
+            $response = $e->getResponse();
+            $message = __('Bad request.');
+            if ($response->getStatusCode() === 401) {
+                $responseBody = json_decode($response->getBody()->getContents(), true);
+                $message = $responseBody['errorMsg'] ?? $response->getReasonPhrase();
+            }
+
+            throw new PublishingFailedException($message, $e->getCode());
+        }
+    }
+
+    private function getPlatformService(): PlatformService
+    {
+        return $this->getServiceLocator()->get(PlatformService::class);
+    }
+
+    /**
+     * @param ResponseInterface $response
+     */
+    private function processApiResponse(ResponseInterface $response): string
+    {
+        if ($response->getStatusCode() !== 200) {
+            throw new PublishingFailedException(__('Request failed with message: %s', $response->getBody()->getContents()));
```

Comment:
It doesn't look like the condition above is ever going to be `true`.
A `ClientException`, transformed into `PublishingFailedException` will be thrown earlier if the response is not a `2xx`, I believe.

Reply:
It is there to handle 500 responses. Previously our `returnFailure` didn't return 200 status code in case of failure response, as I know. It was changed by accident when introduced PSR request/response interfaces in controller, but I don't know id it was/going to be fixed.